frontend/src/App.js

```diff
@@ -12,66 +11,80 @@ class App extends Component {
     super(props);
     this.state = {
       sql: `SELECT COUNT(*) as num_commits, month, repo_id, committer_email FROM (
-    SELECT MONTH(committer_when) as month, r.id as repo_id, committer_email
-    FROM repositories r
-    INNER JOIN refs ON refs.repository_id = r.id AND refs.name = 'HEAD'
-    INNER JOIN commits c ON YEAR(committer_when) = 2018 AND history_idx(refs.hash, c.hash) >= 0
+  SELECT MONTH(committer_when) as month, r.repository_id as repo_id, committer_email
+  FROM repositories r
+  INNER JOIN refs ON refs.repository_id = r.repository_id AND refs.ref_name = 'HEAD'
+  INNER JOIN commits c ON YEAR(committer_when) = 2018 AND history_idx(refs.commit_hash, c.commit_hash) >= 0
 ) as t
 GROUP BY committer_email, month, repo_id`,
-      response: undefined,
-      loading: false,
-      error: undefined
+      results: new Map()
     };
 
     this.handleTextChange = this.handleTextChange.bind(this);
     this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleRemoveResult = this.handleRemoveResult.bind(this);
+
+    this.uniqueKey = 0;
   }
 
   handleTextChange(text) {
     this.setState({ sql: text });
   }
 
   handleSubmit() {
-    this.setState({
-      response: undefined,
-      loading: true,
-      error: undefined
-    });
+    const { sql } = this.state;
+    const key = ++this.uniqueKey;  // eslint-disable-line
+
+    const loadingResults = new Map(this.state.results);
+    loadingResults.set(key, { sql, loading: true });
+
+    this.setState({ results: loadingResults });
 
     api
-      .query(this.state.sql)
-      .then(json => {
-        this.setState({
-          response: json,
-          loading: false
-        });
+      .query(sql)
+      .then(response => {
+        if (!this.state.results.has(key)) {
+          // Tab was removed, ignore results
+          return;
+        }
+
+        const newResults = new Map(this.state.results);
+        newResults.set(key, { sql, response });
+
+        this.setState({ results: newResults });
       })
       .catch(msgArr => {
-        this.setState({ loading: false, error: msgArr.join('; ') });
+        if (!this.state.results.has(key)) {
+          // Tab was removed, ignore results
+          return;
+        }
+
+        const newResults = new Map(this.state.results);
+        newResults.set(key, { sql, errorMsg: msgArr.join('; ') });
```

Comment:
`then` and `catch` is almost the same but for one part of this line... I'd try to refactor...

Reply:
done in 71c1404